File: templates/python/sandbox_config.py
"""Sandbox configuration with Claude Agent SDK."""
import asyncio
from claude_agent_sdk import query, ClaudeAgentOptions, ResultMessage
from claude_agent_sdk.types import (
    ToolPermissionContext, PermissionResultAllow, PermissionResultDeny,
)


async def basic_sandbox():
    """Basic sandbox — auto-approve bash, allow local network."""
    options = ClaudeAgentOptions(
        permission_mode="bypassPermissions",
        sandbox={
            "enabled": True,
            "autoAllowBashIfSandboxed": True,
            "network": {"allowLocalBinding": True},
        },
    )

    async for msg in query(prompt="Build the project and run tests", options=options):
        if isinstance(msg, ResultMessage):
            if msg.subtype == "success":
                print(msg.result)
            else:
                print(f"Error: {msg.subtype}")


async def production_sandbox():
    """Production sandbox — restricted network, canUseTool for escape hatch."""

    async def can_use_tool(
        tool_name: str, tool_input: dict, context: ToolPermissionContext
    ) -> PermissionResultAllow | PermissionResultDeny:
        # Approve or deny unsandboxed command requests
        if tool_name == "Bash" and tool_input.get("dangerouslyDisableSandbox"):
            safe_commands = ["npm publish", "docker push"]
            if any(tool_input.get("command", "").startswith(cmd) for cmd in safe_commands):
                return PermissionResultAllow(updated_input=tool_input)
            return PermissionResultDeny(
                message=f"Unsandboxed blocked: {tool_input.get('command')}"
            )
        return PermissionResultAllow(updated_input=tool_input)

    options = ClaudeAgentOptions(
        can_use_tool=can_use_tool,
        sandbox={
            "enabled": True,
            "autoAllowBashIfSandboxed": True,
            # Static allowlist: always bypass sandbox
            "excludedCommands": ["git", "docker", "kubectl"],
            # Model can request unsandboxed execution — falls back to can_use_tool
            "allowUnsandboxedCommands": True,
            "network": {
                "allowLocalBinding": True,
                "allowUnixSockets": ["/var/run/docker.sock"],
            },
        },
    )

    async for msg in query(prompt="Deploy and verify health checks", options=options):
        if isinstance(msg, ResultMessage) and msg.subtype == "success":
            print(msg.result)


asyncio.run(basic_sandbox())

File: templates/typescript/sandbox-config.ts
import { query } from "@anthropic-ai/claude-agent-sdk";

/**
 * Sandbox Configuration Template
 *
 * Demonstrates:
 * - SandboxSettings with network config
 * - excludedCommands for static allowlist
 * - allowUnsandboxedCommands with canUseTool fallback
 * - Domain-restricted network access
 */

// Basic sandbox — auto-approve bash, allow local network
async function basicSandbox() {
  for await (const message of query({
    prompt: "Build the project and run tests",
    options: {
      permissionMode: "bypassPermissions",
      allowDangerouslySkipPermissions: true,
      sandbox: {
        enabled: true,
        autoAllowBashIfSandboxed: true,
        network: {
          allowLocalBinding: true,
        },
      },
    },
  })) {
    if (message.type === "result") {
      if (message.subtype === "success") {
        console.log(message.result);
      } else {
        console.error(message.subtype, message.errors);
      }
    }
  }
}

// Production sandbox — restricted network, canUseTool for escape hatch
async function productionSandbox() {
  for await (const message of query({
    prompt: "Deploy the application and verify health checks",
    options: {
      sandbox: {
        enabled: true,
        autoAllowBashIfSandboxed: true,
        // Static allowlist: these commands always bypass sandbox
        excludedCommands: ["git", "docker", "kubectl"],
        // Model can request unsandboxed execution, falls back to canUseTool
        allowUnsandboxedCommands: true,
        network: {
          allowLocalBinding: true,
          allowedDomains: ["api.github.com", "registry.npmjs.org"],
          allowManagedDomainsOnly: true,
        },
      },

      // Approve or deny unsandboxed command requests
      canUseTool: async (toolName, input, { signal }) => {
        if (toolName === "Bash" && input.dangerouslyDisableSandbox) {
          const safeCommands = ["npm publish", "docker push"];
          if (safeCommands.some((cmd) => input.command?.startsWith(cmd))) {
            return { behavior: "allow", updatedInput: input };
          }
          return { behavior: "deny", message: `Unsandboxed command blocked: ${input.command}` };
        }
        return { behavior: "allow", updatedInput: input };
      },
    },
  })) {
    if (message.type === "result" && message.subtype === "success") {
      console.log(message.result);
    }
  }
}

basicSandbox().catch(console.error);
